designs/drift.md

```diff
@@ -0,0 +1,81 @@
+|----------------------------|--------|---------|------------|
+| - Provisioner Fields -     |   ---  |   ---   |     ---    |
+| Startup Taints             |    x   |         |            |
+| Taints                     |    x   |         |            |
+| Labels                     |    x   |         |            |
+| Annotations                |    x   |         |            |
+| Node Requirements          |        |    x    |            |
+| Kubelet Configuration      |    x   |         |            |
+| Weight                     |        |         |      x     |
+| Limits                     |        |         |      x     |
+| Consolidation              |        |         |      x     |
+| TTLSecondsUntilExpired     |        |         |      x     |
+| TTLSecondsAfterEmpty       |        |         |      x     |
+| - AWSNodeTemplate Fields - |   ---  |   ---   |     ---    |
+| Subnet Selector            |        |    x    |            |
+| Security Group Selector    |        |    x    |            |
+| Instance Profile           |    x   |         |            |
+| AMI Family/AMI Selector    |        |    x    |            |
+| UserData                   |    x   |         |            |
+| Tags                       |    x   |         |            |
+| Metadata Options           |    x   |         |            |
+| Block Device Mappings      |    x   |         |            |
+| Detailed Monitoring        |    x   |         |            |
+|                            |        |         |            |
+```
+
+Design Questions
+
+## 🔑 API Choices
+
+Drift is an existing mechanism in Karpenter and relies on standard Provisioning CRDs in Karpenter - Provisioners and AWSNodeTemplates. Drift is toggled through a feature gate in the `karpenter-global-settings` ConfigMap.
+
+Once Drift is implemented as described above, the Drift feature gate will be removed, and Drift will be enabled by default. There will be no API additions.
+
+In the future, users may want Karpenter to not Drift certain fields. This promotes users to not rely on CRDs as declarative sources of truth, breaking the contract that Drift defines. If a user request to opt out a field from Drift is validated, Karpenter should first rely on other Deprovisioning control mechanisms to enable this. Otherwise, Karpenter could consider adding a setting in the `karpenter-global-settings` ConfigMap to selectively toggle this.
+
+## 🔑 Static Field Flexibility
+
+### Problem
+
+Users that run other controllers in their clusters that edit machines directly like Cilium and NVIDIA GPU Operator may find Drift too unstable for Static Fields, since Karpenter will drift machines that don’t exactly match the CRDs. If a user has an external controller that edit labels on the machine, or edit tags on the instance, Karpenter will drift those machines any time this external controller makes changes. Karpenter should work in tandem with external systems that users run in their clusters.
```

Comment:
Cilium and NVIDIA GPU are modifying the node not the machine. Does the below edit make sense?

```suggestion
Users that run other controllers in their clusters that edit machines or nodes directly, like Cilium and NVIDIA GPU Operator, may find Drift too unstable for Static Fields, since Karpenter will drift machines that don’t exactly match the CRDs. If a user has an external controller that edit labels on the machines or nodes, or edit tags on the instance, Karpenter will drift those machines any time this external controller makes changes. Karpenter should work in tandem with external systems that users run in their clusters.
```

Reply:
Well in fact they modify both machines and nodes, as you've added (assuming machine means "VM's or bare metal hardware's OS configuration" as distinct from something in the k8s abstraction layer like a node). So your edit above does make sense but these controllers do do stuff that modifies abstractions underneath nodes, to be clear.

Reply:
I think this is one problem with the "Machine" name. This doc is referring to Machine as the `machines.karpenter.sh` CRD.  Things at the VM level are noted as "VM" or "instance".  The edit was meant to signal editing either one of those resources could result in drift consequences.  I wouldn't expect users to take a dependency on the Machine CRD (at least very hopeful they don't).